movement/io/validators.py

```diff
@@ -369,3 +369,148 @@ def __attrs_post_init__(self):
                 "Keypoint names were not provided. "
                 f"Setting to {self.keypoint_names}."
             )
+
+
+@define(kw_only=True)
+class ValidBboxesDataset:
+    """Class for validating bounding boxes data for a ``movement`` dataset.
+
+    We consider 2D bounding boxes only.
+
+    Attributes
+    ----------
+    position_array : np.ndarray
+        Array of shape (n_frames, n_unique_individual_names, n_space)
+        containing the bounding boxes' centroid positions. It will be
+        converted to a `xarray.DataArray` object named "position".
+    shape_array : np.ndarray
+        Array of shape (n_frames, n_unique_individual_names, n_space)
+        containing the bounding boxes' width (extension along the x-axis) and
+        height (extension along the y-axis). It will be converted to a
+        `xarray.DataArray` object named "shape".
+    confidence_array : np.ndarray, optional
+        Array of shape (n_frames, n_individuals, n_keypoints) containing
+        the bounding boxes confidence scores. It will be converted to a
+        `xarray.DataArray` object named "confidence". If None (default), the
+        confidence scores will be set to an array of NaNs.
+    individual_names : list of str, optional
+        List of individual_names for the tracked bounding boxes in the video.
+        If None (default), bounding boxes are assigned unique IDs per frame,
```

Comment:
> In my mind, "unique ID per frame" means that the uniqueness is only enforced within each frame, i.e. in a new frame you'd restart with "id_1". 

I mean exactly that 😬  Usually if there are is no explicit tracking across frames, the detections are simply numbered per frame. I think this makes sense since it makes each frame "independent" - otherwise the numbering of one frame depends on previous frames. 

Note that your dataset may not necessarily be contiguous frames in a video (this is often the case if you load human-labelled data).  With unique IDs per frame, if the data loaded is not tracked, the ID simply refers to the the number of detections on that frame. 

It also keeps numbers manageable. And if there is only one individual, the default behaviour will label it correctly as the same one. I guess the downside is that a user may interpret IDs as being contiguous, which won't happen if the are unique per video - but I was hoping the warning would be sufficient to make them aware of this.

Reply:
So far in `movement` we have always assumed that people are loading data after tracking, and we basically assume that `n_tracks` = `n_individuals`. This is what we do for SLEAP, and we [explicitly state that in the notes](https://movement.neuroinformatics.dev/api/movement.io.load_poses.from_sleap_file.html):

> movement expects the tracks to be assigned and proofread before loading them, meaning each track is interpreted as a single individual. If no tracks are found in the file, movement assumes that this is a single-individual track, and will assign a default individual name.

In most cases, when importing data from a file, the loader will read the track IDs and assign those as `individual_names`. That's what we do for SLEAP, and I assumed we could do the same for the crabs (since we have track IDs for those). Leaving `individual_names=None` essentially is there for the case where you don't know the track IDs (or you don't care about them) and you want `movement` to auto-assign one ID per track. But in all cases, we assume a known number of tracks, which is equal to the `n_individuals`  dimension of the supplied position array. So, I never interpreted `individual_names=None` to mean "no tracking done".

I get your desire to make the class more general, so it can handle un-tracked data, or even non-contiguous labelled data (see #146). I think that is a worthy goal, but no other part of `movement` handles such data as of now, so it would be weird if the bbox object was the only one to achieve this generalisation.

My personal preference would be for the first implementation of this class to behave similarly to `ValidPosesDataset`, aka assume contiguous tracked data. After that we would have to separately think about how to handle untracked or labelled data, but that discussion concerns the entirety of movement, not just the bounding boxes.

I'm not sure if I'm being clear, and I may have misunderstood something you said, so let's have a discussion in person if there's still ambiguity.

Reply:
> I think the confusion comes from the fact that the output from multianimal SLEAP (or DLC) is always tracked. So keypoints across frames are always linked. This means that implicitly track IDs always exist and setting individual_names is just a "cosmetic" choice. There is always some sort of IDs linking keypoints across frames, individual_names just allows you to name these tracks differently. Is this correct?

> So when individual_names=None is set, this doesn't mean that the tracks don't exist (or that you don't know them). It just means that you want them to be named with the default movement convention. But the tracks exist (they always exist for a SLEAP file). Does that make sense? (and is this correct?)

That's exactly it, you've hit the nail on the head.

Reply:
> Thinking about this now, I feel this behaviour may be a bit confusing. I would maybe expect that by default (if individual_names is None), whatever individual names that are specified in the SLEAP/DLC/input file are used. If individual_names is set to something else, then the track names are overwritten with whatever is provided.
>
> Right now, to use the individual names specified in the SLEAP file, you need to pass them explicitly right? I may be off in some of these things, let me know.

Not quite. The thing is that the user will never use these validator objects directly, instead they will use one of the functions that load data from file, e.g. `from_sleap_file()`. Those functions do not expose `individual_names`, because they automatically read whatever track IDs are present and assign those as individual names. The `individual_names=None` option is only really relevant if one loads that data directly from `numpy` arrays, using the `from_numpy()` functions. In that case, they have to explicitly supply the names for individuals and keypoints, because there is no file in which they are stored. So if they don't care about the ID names, of if they forget to pass them, `movement` will auto-assign the default IDs. But this is kind of an edger case, when reading from file, we always use whatever the source software tells us.

Reply:
Btw that's why the title of our posters (so far) was "A Python toolbox for analysing pose **tracks**", because the whole thing was designed around the idea that you already have the tracks (which for DLC and SLEAP is the case).

Reply:
right, thanks for clarifying.

In that case I think the implementation as it is here is consistent with the poses dataset validator. One could call `from_numpy` to create a `ValidBboxesDataset` with `individual_names=None`, in which case the shape of the position array is used to derive the tracks' IDs. As a result, the IDs would be unique per frame (not unique across the whole video).

However for a VIA file this will never happen (this was confusing me). Because in the VIA file validator (PR in progress) we check that tracks have IDs. I think this is good for now.

I'm gonna re-request a review, hopefully I covered all your points. Thanks for the help!

Reply:
also added an [issue](https://github.com/neuroinformatics-unit/movement/issues/220) for supporting untracked data